// particle-usb/lib/particle-usb.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.getDevices = getDevices;
exports.openDeviceById = openDeviceById;

_Object$defineProperty(exports, "DeviceType", {
  enumerable: true,
  get: function () {
    return _deviceType.DeviceType;
  }
});

_Object$defineProperty(exports, "PollingPolicy", {
  enumerable: true,
  get: function () {
    return _deviceBase.PollingPolicy;
  }
});

_Object$defineProperty(exports, "FirmwareModule", {
  enumerable: true,
  get: function () {
    return _device.FirmwareModule;
  }
});

_Object$defineProperty(exports, "WifiAntenna", {
  enumerable: true,
  get: function () {
    return _wifiDevice.WifiAntenna;
  }
});

_Object$defineProperty(exports, "WifiSecurity", {
  enumerable: true,
  get: function () {
    return _wifiDevice.WifiSecurity;
  }
});

_Object$defineProperty(exports, "WifiCipher", {
  enumerable: true,
  get: function () {
    return _wifiDevice.WifiCipher;
  }
});

_Object$defineProperty(exports, "EapMethod", {
  enumerable: true,
  get: function () {
    return _wifiDevice.EapMethod;
  }
});

_Object$defineProperty(exports, "CloudConnectionStatus", {
  enumerable: true,
  get: function () {
    return _cloudDevice.CloudConnectionStatus;
  }
});

_Object$defineProperty(exports, "ServerProtocol", {
  enumerable: true,
  get: function () {
    return _cloudDevice.ServerProtocol;
  }
});

_Object$defineProperty(exports, "NetworkStatus", {
  enumerable: true,
  get: function () {
    return _networkDevice.NetworkStatus;
  }
});

_Object$defineProperty(exports, "Result", {
  enumerable: true,
  get: function () {
    return _result.Result;
  }
});

_Object$defineProperty(exports, "DeviceError", {
  enumerable: true,
  get: function () {
    return _error.DeviceError;
  }
});

_Object$defineProperty(exports, "NotFoundError", {
  enumerable: true,
  get: function () {
    return _error.NotFoundError;
  }
});

_Object$defineProperty(exports, "NotAllowedError", {
  enumerable: true,
  get: function () {
    return _error.NotAllowedError;
  }
});

_Object$defineProperty(exports, "StateError", {
  enumerable: true,
  get: function () {
    return _error.StateError;
  }
});

_Object$defineProperty(exports, "TimeoutError", {
  enumerable: true,
  get: function () {
    return _error.TimeoutError;
  }
});

_Object$defineProperty(exports, "MemoryError", {
  enumerable: true,
  get: function () {
    return _error.MemoryError;
  }
});

_Object$defineProperty(exports, "ProtocolError", {
  enumerable: true,
  get: function () {
    return _error.ProtocolError;
  }
});

_Object$defineProperty(exports, "UsbError", {
  enumerable: true,
  get: function () {
    return _error.UsbError;
  }
});

_Object$defineProperty(exports, "InternalError", {
  enumerable: true,
  get: function () {
    return _error.InternalError;
  }
});

_Object$defineProperty(exports, "RequestError", {
  enumerable: true,
  get: function () {
    return _error.RequestError;
  }
});

_Object$defineProperty(exports, "config", {
  enumerable: true,
  get: function () {
    return _config.config;
  }
});

exports.AssetTracker = exports.B5Som = exports.XenonSom = exports.BoronSom = exports.ArgonSom = exports.Xenon = exports.Boron = exports.Argon = exports.Electron = exports.P1 = exports.Photon = exports.Core = void 0;

var _setPrototypeOf = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/set-prototype-of"));

var _deviceType = require("./device-type");

var _deviceBase = require("./device-base");

var _device = require("./device");

var _wifiDevice = require("./wifi-device");

var _cellularDevice = require("./cellular-device");

var _cloudDevice = require("./cloud-device");

var _meshDevice = require("./mesh-device");

var _networkDevice = require("./network-device");

var _result = require("./result");

var _error = require("./error");

var _config = require("./config");

class Core extends _deviceBase.DeviceBase {}

exports.Core = Core;

class Photon extends (0, _cloudDevice.CloudDevice)((0, _wifiDevice.WifiDevice)((0, _networkDevice.NetworkDevice)(_device.Device))) {}

exports.Photon = Photon;

class P1 extends (0, _cloudDevice.CloudDevice)((0, _wifiDevice.WifiDevice)((0, _networkDevice.NetworkDevice)(_device.Device))) {}

exports.P1 = P1;

class Electron extends (0, _cloudDevice.CloudDevice)((0, _cellularDevice.CellularDevice)((0, _networkDevice.NetworkDevice)(_device.Device))) {}

exports.Electron = Electron;

class Argon extends (0, _cloudDevice.CloudDevice)((0, _wifiDevice.WifiDevice)((0, _meshDevice.MeshDevice)((0, _networkDevice.NetworkDevice)(_device.Device)))) {}

exports.Argon = Argon;

class Boron extends (0, _cloudDevice.CloudDevice)((0, _cellularDevice.CellularDevice)((0, _meshDevice.MeshDevice)((0, _networkDevice.NetworkDevice)(_device.Device)))) {}

exports.Boron = Boron;

class Xenon extends (0, _cloudDevice.CloudDevice)((0, _meshDevice.MeshDevice)((0, _networkDevice.NetworkDevice)(_device.Device))) {}

exports.Xenon = Xenon;

class ArgonSom extends (0, _cloudDevice.CloudDevice)((0, _wifiDevice.WifiDevice)((0, _meshDevice.MeshDevice)((0, _networkDevice.NetworkDevice)(_device.Device)))) {}

exports.ArgonSom = ArgonSom;

class BoronSom extends (0, _cloudDevice.CloudDevice)((0, _cellularDevice.CellularDevice)((0, _meshDevice.MeshDevice)((0, _networkDevice.NetworkDevice)(_device.Device)))) {}

exports.BoronSom = BoronSom;

class XenonSom extends (0, _cloudDevice.CloudDevice)((0, _meshDevice.MeshDevice)((0, _networkDevice.NetworkDevice)(_device.Device))) {}

exports.XenonSom = XenonSom;

class B5Som extends (0, _cloudDevice.CloudDevice)((0, _cellularDevice.CellularDevice)((0, _meshDevice.MeshDevice)((0, _networkDevice.NetworkDevice)(_device.Device)))) {}

exports.B5Som = B5Som;

class AssetTracker extends (0, _cloudDevice.CloudDevice)((0, _cellularDevice.CellularDevice)((0, _networkDevice.NetworkDevice)(_device.Device))) {}

exports.AssetTracker = AssetTracker;
const DEVICE_PROTOTYPES = {
  [_deviceType.DeviceType.CORE]: Core.prototype,
  [_deviceType.DeviceType.PHOTON]: Photon.prototype,
  [_deviceType.DeviceType.P1]: P1.prototype,
  [_deviceType.DeviceType.ELECTRON]: Electron.prototype,
  [_deviceType.DeviceType.ARGON]: Argon.prototype,
  [_deviceType.DeviceType.BORON]: Boron.prototype,
  [_deviceType.DeviceType.XENON]: Xenon.prototype,
  [_deviceType.DeviceType.ARGON_SOM]: ArgonSom.prototype,
  [_deviceType.DeviceType.BORON_SOM]: BoronSom.prototype,
  [_deviceType.DeviceType.XENON_SOM]: XenonSom.prototype,
  [_deviceType.DeviceType.B5_SOM]: B5Som.prototype,
  [_deviceType.DeviceType.ASSET_TRACKER]: AssetTracker.prototype
};

function setDevicePrototype(dev) {
  const proto = DEVICE_PROTOTYPES[dev.type];

  if (!proto) {
    return dev;
  }

  return (0, _setPrototypeOf.default)(dev, proto);
}

function getDevices(options) {
  return (0, _deviceBase.getDevices)(options).then(devs => devs.map(dev => setDevicePrototype(dev)));
}

function openDeviceById(id, options) {
  return (0, _deviceBase.openDeviceById)(id, options).then(dev => setDevicePrototype(dev));
}
